### ISSUE TITLE
Question About Loss

### ISSUE THREAD

**Issue by shuxiaobo** [seal-rg/recurrent-pretraining#19]
hi there, awsome work you have done, recurrent is a noval idea to improve model reasoning perform
I hava a question about the loss declare at `3.3 Training Objective`, here is :`we randomly sample iteration counts during training, assigning a random number of iterations r to every input sequence`, dose every sequence means every batch when training?

**Comment by shuxiaobo** [comment]
I've found ``

**Comment by JonasGeiping** [comment]
A self-learned/dynamic  r would be a "halting module" in the language of the universal transformers papers (https://arxiv.org/abs/1807.03819). We found this a bit hard to get right at scale - the model has to simultaneously learn when to exit or extend computation, and learn to solve next-token-prediction as well as it can.
